### PR TITLE
Add Bot read API endpoints

### PR DESCRIPTION
**Pull request by miller46**
## Summary
- Add GET /api/v1/bots endpoint with pagination, filtering, and sorting
- Add GET /api/v1/bots/{id} endpoint for single bot retrieval
- Add database session dependency for FastAPI
- Add Pydantic response schemas (BotResponse, BotListResponse)
- Fix Bot model `last_run_at` field default value

## Test plan
- [x] All 28 unit tests pass
- [x] Ruff linting passes
- [ ] Manual testing of endpoints with sample data
- [ ] Verify pagination works correctly with large datasets
- [ ] Verify all filter combinations work as expected

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/jm_api/api/routes/bots.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    # Get total count
+    count_query = select(func.count()).select_from(query.subquery())
+    total = db.execute(count_query).scalar() or 0
+
+    # Apply sorting and pagination
+    query = query.order_by(Bot.create_at.desc())
+    offset = (page - 1) * per_page
+    query = query.offset(offset).limit(per_page)
+
+    # Execute query
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: **Nit: `pages` calculation edge case** — When `total > 0`, `pages` should always be at least 1, and `math.ceil` handles this. But consider that if `page > pages`, the response returns an empty `items` list with misleading pagination metadata (e.g. `page=999, pages=2`). Consider either clamping `page` to `pages`, or returning a 404/422 when `page > pages`.

File: src/jm_api/schemas/bot.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    rig_id: str
+    last_run_at: datetime | None
+    kill_switch: bool
+    last_run_log: str | None
+    create_at: datetime
+    last_update_at: datetime
+
+    model_config = ConfigDict(
+        from_attributes=True,
+        json_schema_extra={
```
Comment (miller46bot) [inline src/jm_api/schemas/bot.py]: **Nit:** The `json_schema_extra` example uses `False` (Python bool) instead of `false` (JSON). Pydantic handles this, but it reads oddly since this is nominally JSON schema content. Minor.

File: src/jm_api/models/bot.py [hunk 3]
```diff
@@ -12,6 +12,6 @@ class Bot(TimestampedIdBase):
     __tablename__ = "bots"
 
     rig_id: Mapped[str] = mapped_column(String(128), nullable=False)
-    last_run_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
+    last_run_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), default=None)
     kill_switch: Mapped[bool] = mapped_column(Boolean, default=False)
     last_run_log: Mapped[str | None] = mapped_column(Text, default="")
```
Comment (miller46bot) [inline src/jm_api/models/bot.py]: **`last_run_log` default is `""` (empty string) not `None`.** The `Mapped` type says `str | None` but `default=""` means new bots get an empty string rather than `None`. The schema (`BotResponse`) declares `last_run_log: str | None`. This inconsistency means the API will never actually return `None` for new bots — it'll return `""`. Decide on one: if `None` means "never ran" and `""` means "ran with no output", document it. Otherwise, align the default with the intent.

File: src/jm_api/schemas/bot.py (hunk 2 above)
Comment (miller46bot) [inline src/jm_api/schemas/bot.py]: **`json_schema_extra` example uses a hardcoded date in 2024.** This is cosmetic, but consider whether the example date matters for documentation. It's fine as-is but will look stale over time.

File: tests/test_bots.py [hunk 4]
```diff
@@ -0,0 +1,628 @@
+"""Tests for Bot read API endpoints."""
```
Comment (miller46bot) [inline tests/test_bots.py]: **Test file is comprehensive — nice coverage.** 28 tests covering empty lists, pagination, filters, date ranges, combined filters, sorting, 404s, SQL injection prevention, and deterministic ordering. Well structured with clear class groupings.

File: tests/test_bots.py [hunk 5]
```diff
@@ -0,0 +1,628 @@
...
+    """Create database session from test engine."""
+    session = Session(db_engine)
+    yield session
+    session.close()
+
+
+@pytest.fixture
+def app(db_session: Session) -> FastAPI:
+    """Create test app with overridden database dependency."""
+    from jm_api.app import create_app
```
Comment (miller46bot) [inline tests/test_bots.py]: **Test fixtures duplicate setup across test files.** `db_engine`, `db_session`, and `app` fixtures are defined here but `test_models.py` has its own `setup_in_memory_db()`. Consider moving shared DB fixtures to a `conftest.py` to avoid duplication and ensure consistent test setup.

File: tests/test_bots.py [hunk 6]
```diff
@@ -0,0 +1,628 @@
...
+    return TestClient(app)
+
+
+def create_bot(
+    session: Session,
+    rig_id: str = "rig-001",
+    kill_switch: bool = False,
+    last_run_log: str | None = None,
+    last_run_at: datetime | None = None,
+) -> Bot:
```
Comment (miller46bot) [inline tests/test_bots.py]: **`create_bot` helper doesn't allow setting `create_at`.** Several date filter tests (e.g., `test_filter_create_at_after`) rely on insertion order and timing to differentiate `create_at` values. This is fragile — if both bots are created within the same clock tick, the tests could flake. Consider allowing `create_at` to be passed explicitly (or use `freezegun`/manual timestamp injection) for deterministic date-based tests.

File: tests/test_bots.py [hunk 7]
```diff
@@ -0,0 +1,628 @@
...
+    ) -> None:
+        """log_search matches substrings."""
+        # Arrange
+        create_bot(db_session, rig_id="rig-001", last_run_log="Task failed")
+        create_bot(db_session, rig_id="rig-002", last_run_log="failure detected")
+        create_bot(db_session, rig_id="rig-003", last_run_log="Success")
+
+        # Act
+        response = client.get("/api/v1/bots", params={"log_search": "fail"})
+
```
Comment (miller46bot) [inline tests/test_bots.py]: **Date filter test assertions are weak.** `assert data["total"] >= 1` is a very loose assertion — it doesn't verify that bots *outside* the range are excluded. In these tests, both bots might have the same `create_at` (same clock tick in SQLite), making the filter effectively a no-op. The tests pass but don't actually prove the filter works. Use explicit timestamps to make these tests deterministic.

File: src/jm_api/core/config.py [hunk 8]
```diff
@@ -12,6 +12,8 @@ class Settings(BaseSettings):
     environment: str = Field(default="development")
     debug: bool = Field(default=False)
 
+    database_url: str = Field(default="sqlite:///./app.db")
```
Comment (miller46bot) [inline src/jm_api/core/config.py]: **SQLite default for `database_url` is fine for development** but `sqlite:///./app.db` will create the DB relative to the working directory, which can be surprising. Consider using an absolute path or documenting this behavior.

File: tests/test_models.py [hunk 9]
```diff
@@ -31,15 +31,18 @@ def test_bot_id_generated_and_format() -> None:
 def test_bot_timestamps_create_and_update() -> None:
     session = setup_in_memory_db()
```
Comment (miller46bot) [inline tests/test_models.py]: **Good fix for SQLite timezone handling.** Replacing UTC timezone assertions with before/after bounds is the right approach for SQLite compatibility. The `.replace(tzinfo=None)` stripping is a pragmatic workaround.

**Review by miller46bot — CHANGES_REQUESTED** [review]
Overall a solid PR with good test coverage and clean separation of concerns. There are a few issues to address ranging from a bug in the error response to design improvements.

File: src/jm_api/api/routes/bots.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+
+@router.get(
+    "/{bot_id}",
+    response_model=BotResponse,
+    responses={404: {"model": BotNotFoundError}},
+)
+def get_bot(bot_id: str, db: Session = Depends(get_db)) -> BotResponse:
+    """Retrieve a single bot by ID."""
+    bot = db.get(Bot, bot_id)
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: Bug: The `detail` kwarg to `HTTPException` IS the value that FastAPI serializes under the `"detail"` key in the response JSON. By passing `{"detail": "Bot not found", "id": bot_id}`, the response body becomes `{"detail": {"detail": "Bot not found", "id": "..."}}` - a double-nested structure. The test on line 460 of test_bots.py confirms this awkward nesting (`data["detail"]["detail"]`). Either:
- Pass a flat string: `detail=f"Bot {bot_id} not found"` and update the test, or
- Pass `detail={"message": "Bot not found", "id": bot_id}` to avoid the key name collision.

File: src/jm_api/schemas/bot.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+class BotListResponse(BaseModel):
+    """Paginated list of bots response schema."""
+
+    items: list[BotResponse]
+    total: int
+    page: int
+    per_page: int
+    pages: int
+
```
Comment (miller46bot) [inline src/jm_api/schemas/bot.py]: The `BotNotFoundError` schema declares `detail: str` and `id: str` as top-level fields, but the actual HTTPException response nests these under a `detail` key (i.e. `{"detail": {"detail": ..., "id": ...}}`). This schema will never match the real response shape, so the `responses={404: {"model": BotNotFoundError}}` annotation in the route generates misleading OpenAPI docs. Either fix the error format (see other comment) or wrap the schema to match reality.

File: src/jm_api/api/routes/bots.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+from jm_api.db.session import get_db
+from jm_api.models.bot import Bot
+from jm_api.schemas.bot import BotListResponse, BotNotFoundError, BotResponse
+
+router = APIRouter(prefix="/bots", tags=["bots"])
+
+
+@router.get("", response_model=BotListResponse)
+def list_bots(
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: The return type annotation is `-> BotListResponse` but FastAPI's `response_model=BotListResponse` handles serialization. Since `list_bots` actually returns a `BotListResponse` Pydantic model directly, this works, but note that `get_bot` (line 86) returns `BotResponse.model_validate(bot)` which is a Pydantic model, not a SQLAlchemy model. Consider returning the ORM object directly and letting `response_model` handle validation/serialization - this avoids double validation and is the idiomatic FastAPI pattern.

**Review by miller46bot — CHANGES_REQUESTED** [review]
Generally solid PR that adds well-structured read endpoints with good test coverage. The main issues are: a bug in the 404 error response structure (double-nested `detail`), the `BotNotFoundError` schema not matching actual error output, a return type mismatch in the route handlers, and `lru_cache` on engine/session factory making test teardown fragile. Several smaller improvements around API naming conventions and test robustness are also noted.

**Review by miller46bot — CHANGES_REQUESTED** [review]
Solid PR overall — good test coverage (28 tests), proper SQL wildcard escaping, deterministic sorting, and clean separation of concerns. A few issues to address before merging:

**Bugs / Correctness:**

1. **404 response body is double-nested** — `HTTPException(detail={"detail": "Bot not found", "id": bot_id})` produces `{"detail": {"detail": "Bot not found", "id": "..."}}`. The outer `detail` key is added by FastAPI automatically. Should be `detail=f"Bot not found: {bot_id}"` or a flat dict like `{"message": "Bot not found", "id": bot_id}`. The test asserts the double-nested form (`data["detail"]["detail"]`), masking the bug.

2. **`BotNotFoundError` schema is dead code** — It's defined in `schemas/bot.py` and referenced in `responses={404: ...}` for OpenAPI docs, but the actual `HTTPException` produces a different shape. The schema and response will never match.

3. **`list_bots` return type annotation lies** — The function signature says `-> BotListResponse` but FastAPI's `response_model=BotListResponse` handles serialization. Since `BotListResponse` is already constructed and returned, this is fine at runtime, but the function also returns the Pydantic model directly so it works. However, `get_bot` returns `BotResponse.model_validate(bot)` with return type `-> BotResponse`, which is correct, but the `response_model` param is redundant when the return type already specifies it. Minor, not blocking.

**Design / Code Quality:**

4. **`get_engine` / `get_session_factory` use `@lru_cache` on module-level functions** — This works but creates hidden global state that's hard to manage in testing. The `reset_db_state()` function exists to clear it, but it's never called in the test fixtures. The tests work because they override `get_db` directly, but this leaky cached state could cause issues if tests ever exercise the real dependency path.

5. **Pagination `pages` calculation off-by-one on boundary** — `pages=0` when `total=0` is correct, but requesting `page=2` on a 5-item dataset with `per_page=20` (which has `pages=1`) returns 200 with empty items. Consider returning 404 or at minimum documenting this as intentional. This is a design choice, not a bug.

6. **`per_page` cap is silent** — When a client sends `per_page=200`, the response shows `per_page=100` with no indication the value was capped. This is a reasonable default but could surprise API consumers. Consider documenting this behavior in the endpoint docstring or returning a header.

**Nits / Minor:**

7. **Unused `datetime` import in `bots.py`** — `from datetime import datetime` is imported but `datetime` is only used as a type annotation in the query parameter signatures. FastAPI handles parsing from the `Query` annotation. This import is actually needed for the type hints, so this is fine — ignore.

8. **`conftest.py` `create_bot` helper does commit + refresh in both branches** — The `if/else` on `create_at` both call `session.commit()` and `session.refresh(bot)`. Could deduplicate, but it's test code and readability is fine.

9. **Test imports use relative `from conftest import create_bot`** — This works with pytest's conftest magic but is unconventional. Standard practice is to put helpers directly as fixtures or use `pytest` plugin patterns. Not blocking.

File: src/jm_api/api/routes/bots.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    return BotListResponse(
+        items=[BotResponse.model_validate(bot) for bot in bots],
+        total=total,
+        page=page,
+        per_page=per_page,
+        pages=pages,
+    )
+
+
+@router.get(
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: Bug: The 404 response body is double-nested. `HTTPException(detail=...)` wraps `detail` in a `{"detail": ...}` key automatically. So the client receives `{"detail": {"detail": "Bot not found", "id": "..."}}`. The test on line 451 of test_bots.py confirms this awkward structure (`data["detail"]["detail"]`). Either pass a plain string `detail="Bot not found"` and return the id separately, or use a flat dict like `detail={"message": "Bot not found", "id": bot_id}` to avoid the confusing double-`detail` nesting.

File: src/jm_api/models/bot.py (hunk 3 above)
Comment (miller46bot) [inline src/jm_api/models/bot.py]: The change `default=None` for `last_run_at` is correct since `Mapped[datetime | None]` already implies nullability. However, `last_run_log` on line 18 has `default=""` — an empty string as default for a `Mapped[str | None]` field is inconsistent. It means a bot that has never run will have `last_run_log=""` rather than `None`. Consider whether `None` would be more semantically correct for 'never ran'.

File: tests/conftest.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
+"""Shared test fixtures."""
```
Comment (miller46bot) [inline tests/conftest.py]: Good test architecture — in-memory SQLite with StaticPool, dependency override for FastAPI, and a clean `create_bot` helper. One concern: `conftest.py` is in `tests/` but `test_bots.py` imports from it with `from conftest import create_bot` (line 8 of test_bots.py). This works because pytest adds conftest to the path, but it's fragile. Consider making `create_bot` a pytest fixture instead, or using a relative import pattern.

File: src/jm_api/api/routes/bots.py [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+        query = query.where(Bot.last_run_at >= last_run_at_after)
+    if last_run_at_before is not None:
+        query = query.where(Bot.last_run_at <= last_run_at_before)
+
+    # Get total count
+    count_query = select(func.count()).select_from(query.subquery())
+    total = db.execute(count_query).scalar() or 0
+
+    # Apply sorting and pagination (secondary sort by id for deterministic order)
+    query = query.order_by(Bot.create_at.desc(), Bot.id.desc())
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: The `pages` calculation returns `0` when there are no results. This means `page=1` with `pages=0` in an empty response, which is slightly odd — `page` exceeds `pages`. Consider returning `pages=1` when the result is empty (the user is on page 1 of 1 empty pages) or `pages=0` with `page=0`. This is a minor API design decision but worth being intentional about.

File: src/jm_api/models/bot.py [hunk 16]
```diff
@@ -12,6 +12,6 @@ class Bot(TimestampedIdBase):
     __tablename__ = "bots"
 
     rig_id: Mapped[str] = mapped_column(String(128), nullable=False)
-    last_run_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
+    last_run_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), default=None)
```
Comment (miller46bot) [inline src/jm_api/models/bot.py]: Good fix — adding `default=None` to `last_run_at` makes the nullable intent explicit. Without it, SQLAlchemy's `Mapped[datetime | None]` type hint implies nullable but the column definition didn't specify a default, which could cause issues with some DB backends.

File: src/jm_api/api/routes/bots.py [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    pages = math.ceil(total / per_page) if total > 0 else 0
+
+    return BotListResponse(
+        items=[BotResponse.model_validate(bot) for bot in bots],
+        total=total,
+        page=page,
+        per_page=per_page,
+        pages=pages,
+    )
+
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: Bug: The 404 response body is double-nested. `HTTPException(detail=...)` wraps its argument in a `{"detail": ...}` key automatically. So passing `detail={"detail": "Bot not found", "id": bot_id}` results in a response like `{"detail": {"detail": "Bot not found", "id": "..."}}`. The test at test_bots.py:411 confirms this by asserting `data["detail"]["detail"]`, which is awkward for API consumers.

Either pass a plain string: `detail=f"Bot {bot_id} not found"` or pass `detail={"message": "Bot not found", "id": bot_id}` to avoid the double-"detail" nesting.

File: src/jm_api/api/routes/bots.py [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+    offset = (page - 1) * per_page
+    query = query.offset(offset).limit(per_page)
+
+    # Execute query
+    bots = db.execute(query).scalars().all()
+
+    # Calculate pages
+    pages = math.ceil(total / per_page) if total > 0 else 0
+
+    return BotListResponse(
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: The `responses={404: {"model": BotNotFoundError}}` in the decorator doesn't match the actual error shape produced by `HTTPException`. FastAPI's `HTTPException` wraps the detail in `{"detail": ...}`, so the OpenAPI docs will show a schema that doesn't match the real response. The `BotNotFoundError` schema (with top-level `detail` and `id` fields) won't match the actual `{"detail": {"detail": ..., "id": ...}}` response. Consider aligning the schema with reality or using a custom exception handler.

File: src/jm_api/api/routes/bots.py [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+@router.get("", response_model=BotListResponse)
+def list_bots(
+    page: int = Query(default=1, ge=1),
+    per_page: int = Query(default=20, ge=1),
+    rig_id: str | None = Query(default=None),
+    kill_switch: bool | None = Query(default=None),
+    log_search: str | None = Query(default=None),
+    create_at_after: datetime | None = Query(default=None),
+    create_at_before: datetime | None = Query(default=None),
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: Typo in parameter names: `create_at_after` / `create_at_before` — should this be `created_at`? The model field is `create_at` (which is itself unconventional), so it's consistent with the model, but `created_at` is the standard naming. If the model name can't change, this is fine, but worth calling out.

File: src/jm_api/api/routes/bots.py [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+    last_update_at_after: datetime | None = Query(default=None),
+    last_update_at_before: datetime | None = Query(default=None),
+    last_run_at_after: datetime | None = Query(default=None),
+    last_run_at_before: datetime | None = Query(default=None),
+    db: Session = Depends(get_db),
+) -> BotListResponse:
+    """List all bots with filtering, pagination, and sorting."""
+    # Cap per_page at 100
+    per_page = min(per_page, 100)
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: The `per_page = min(per_page, 100)` cap is applied silently — the client asks for 200 but gets 100 without any indication. The response does return `per_page: 100` which helps, but consider whether a 422 error for invalid values would be more appropriate, or at minimum document this behavior in the OpenAPI description.

File: src/jm_api/api/routes/bots.py (hunk 13 above)
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: Bug: The 404 response has double-nested `detail`. FastAPI's `HTTPException` wraps the `detail` kwarg under a `"detail"` key automatically, so the response body becomes `{"detail": {"detail": "Bot not found", "id": "..."}}`. The test on line 441 (`data["detail"]["detail"]`) confirms this double nesting. Change to:
```python
raise HTTPException(
    status_code=404,
    detail=f"Bot {bot_id} not found",
)
```
And update `BotNotFoundError` and the test accordingly. The current schema with a nested `detail` key is confusing for API consumers.

File: src/jm_api/api/routes/bots.py (hunk 18 above)
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: The return type annotation says `-> BotResponse` but the function actually returns a `BotResponse` Pydantic model instance. This works because of `response_model`, but it's slightly misleading. Consider returning the ORM object directly and letting FastAPI + `response_model` handle serialization, or keep it as-is but be aware of the redundancy between `response_model=BotResponse` and the explicit `model_validate` call.

File: src/jm_api/api/routes/bots.py (hunk 12 above)
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: Same comment as above for `list_bots` — the return type is `-> BotListResponse` and you're also specifying `response_model=BotListResponse`. Pick one pattern. The explicit `model_validate` + return type approach bypasses FastAPI's `response_model` serialization (FastAPI sees the object is already the right type). This is fine, but the `response_model` param then only serves for OpenAPI docs generation — worth a brief comment or just drop the return annotation in favor of `response_model`.

File: src/jm_api/db/session.py [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+from functools import lru_cache
+
+from sqlalchemy import create_engine
+from sqlalchemy.engine import Engine
+from sqlalchemy.orm import Session, sessionmaker
+
+from jm_api.core.config import get_settings
+
+
+@lru_cache
```
Comment (miller46bot) [inline src/jm_api/db/session.py]: `@lru_cache` on `get_engine()` and `get_session_factory()` means these are module-level singletons that persist for the process lifetime. This is fine for production, but `reset_db_state()` only clears the cache — it doesn't dispose the existing engine or close open connections. Consider calling `get_engine().dispose()` inside `reset_db_state()` before clearing the cache to avoid connection leaks during testing or config changes.

File: src/jm_api/core/config.py [hunk 22]
```diff
@@ -2,16 +2,33 @@
...
     app_version: str = Field(default="0.1.0")
     environment: str = Field(default="development")
     debug: bool = Field(default=False)
 
+    database_url: str = Field(default="sqlite:///./app.db")
+
+    @model_validator(mode="after")
+    def validate_database_url_for_environment(self) -> "Settings":
+        """Validate database_url is appropriate for the environment."""
+        if self.environment in _PRODUCTION_ENVIRONMENTS:
```
Comment (miller46bot) [inline src/jm_api/core/config.py]: The `database_url` field has a default of `"sqlite:///./app.db"` which creates a file-based database relative to the working directory. For development, an in-memory SQLite (`sqlite:///:memory:`) or a more explicit path would be safer to avoid accidentally creating `app.db` files in unexpected locations when running the app from different directories.

File: src/jm_api/api/routes/bots.py [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+
+    return BotListResponse(
+        items=[BotResponse.model_validate(bot) for bot in bots],
+        total=total,
+        page=page,
+        per_page=per_page,
+        pages=pages,
+    )
+
+
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: Bug: The `get_bot` return type annotation says `-> BotResponse` but the function returns `BotResponse.model_validate(bot)` on the happy path AND raises `HTTPException` on the error path. The return type is fine for the happy path, but the actual issue is that the function signature declares `response_model=BotResponse` on the decorator AND uses `-> BotResponse` as the return annotation. When FastAPI sees both, it double-validates. This is minor but inconsistent with `list_bots` which correctly returns `BotListResponse` (the actual Pydantic model). Both approaches work, just noting the inconsistency.

File: src/jm_api/schemas/bot.py [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+
+class BotListResponse(BaseModel):
+    """Paginated list of bots response schema."""
+
+    items: list[BotResponse]
+    total: int
+    page: int
+    per_page: int
+    pages: int
```
Comment (miller46bot) [inline src/jm_api/schemas/bot.py]: The `BotNotFoundError` schema is declared and referenced in the `responses` dict on the endpoint, but it doesn't match the actual error response shape. HTTPException produces `{"detail": {...}}`, not `{"detail": "...", "id": "..."}`. Either align the schema with the real response shape (wrap it in a container with a `detail` field), or simplify the HTTPException to return a plain string and remove this custom error schema.

File: src/jm_api/api/routes/bots.py (hunk 20 above)
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: Nit: The `per_page` cap is done silently via `per_page = min(per_page, 100)`. This means the response `per_page` field reflects 100 when the client requested 200, but the client has no explicit signal that their value was capped. This is a design choice — it's fine, but consider whether the API contract should reject values > 100 with a 422 (via `le=100` on the Query) rather than silently capping. Explicit rejection is more predictable for API consumers.

File: src/jm_api/db/session.py [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+def get_engine() -> Engine:
+    """Create and cache database engine."""
+    settings = get_settings()
+    return create_engine(settings.database_url)
```
Comment (miller46bot) [inline src/jm_api/db/session.py]: The `@lru_cache` on `get_engine()` means this creates a single global engine that lives forever. This is fine for production, but the `reset_db_state()` function at the bottom suggests you're aware of the testing concern. However, `get_settings()` is also `@lru_cache`'d (in config.py), so if `get_settings` cache isn't cleared too, `reset_db_state` might still get a stale `database_url`. Consider whether `reset_db_state` should also call `get_settings.cache_clear()`.

File: tests/test_bots.py [hunk 26]
```diff
@@ -0,0 +1,574 @@
+"""Tests for Bot read API endpoints."""
+
+from datetime import datetime, timezone
+
+from fastapi.testclient import TestClient
+from sqlalchemy.orm import Session
+
+from conftest import create_bot
```
Comment (miller46bot) [inline tests/test_bots.py]: This import uses `from conftest import create_bot` which works because `tests/` is on the Python path, but it's fragile. If the test directory structure changes or pytest configuration changes, this breaks. The standard pattern is to just use `create_bot` directly in tests since pytest makes conftest fixtures and helpers available. Consider making `create_bot` a fixture that returns a factory function, or import it as `from tests.conftest import create_bot`.

File: tests/conftest.py [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+    app.dependency_overrides[get_db] = override_get_db
+    return app
+
+
+@pytest.fixture
+def client(app: FastAPI) -> TestClient:
+    """Create test client."""
+    return TestClient(app)
+
+
```
Comment (miller46bot) [inline tests/conftest.py]: Good: The `create_bot` helper is well-designed with sensible defaults and the `create_at` override pattern using a raw SQL update is a pragmatic solution for testing time-based filters. One thing to note: `last_update_at` is not overridable, which could matter if you later add filters on `last_update_at_before/after` (which the endpoint already supports). Consider adding a `last_update_at` override parameter for completeness.

File: src/jm_api/models/bot.py (hunk 16 above)
Comment (miller46bot) [inline src/jm_api/models/bot.py]: The change from `mapped_column(DateTime(timezone=True))` to `mapped_column(DateTime(timezone=True), default=None)` — the `default=None` is redundant since `Mapped[datetime | None]` already tells SQLAlchemy the column is nullable and defaults to NULL. Not harmful, but unnecessary.

File: tests/test_bots.py [hunk 28]
```diff
@@ -0,0 +1,574 @@
...
+
+        # Assert
+        assert response.status_code == 200
+        data = response.json()
+        assert len(data["items"]) == 100
+        assert data["per_page"] == 100
+
+    def test_list_bots_per_page_zero_invalid(self, client: TestClient) -> None:
+        """per_page=0 returns 422 validation error."""
+        # Act
```
Comment (miller46bot) [inline tests/test_bots.py]: The test `test_list_bots_per_page_max_capped` creates 150 bots just to verify the cap. This is slow. You could create 2 bots and verify `per_page` in the response is 100 and `len(items)` is 2. The cap logic is independent of the actual row count.

File: src/jm_api/api/routes/bots.py [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+    # Apply filters (AND logic)
+    if rig_id is not None:
+        query = query.where(Bot.rig_id == rig_id)
+    if kill_switch is not None:
+        query = query.where(Bot.kill_switch == kill_switch)
+    if log_search is not None:
+        # Escape SQL wildcards to prevent injection
+        escaped = log_search.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")
+        query = query.where(Bot.last_run_log.ilike(f"%{escaped}%", escape="\\"))
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: The SQL wildcard escaping is good and the tests cover it well. One edge case: if `log_search` is an empty string `""`, the query becomes `WHERE last_run_log ILIKE '%%'` which matches all non-NULL logs. Consider whether empty string should be treated as "no filter" (skip the clause entirely) to be consistent with the `None` behavior.
